Example/ViewController.swift

```diff
@@ -44,13 +50,229 @@ class ViewController: UIViewController {
+
+    private func updateRouteAnnotations(_ routes: [Route]?, style: MGLStyle) {
+        // remove any existing route annotation
+        removeRouteAnnotationsLayerFromStyle(style)
+
+        guard self.waypoints.count > 0, let routes = routes, let mapView = self.mapView else { return }
+
+        let visibleBoundingBox = BoundingBox(coordinateBounds: mapView.visibleCoordinateBounds)
+
+        let tollRoutes = routes.filter { route -> Bool in
+            return (route.tollIntersections?.count ?? 0) > 0
+        }
+        let routesContainTolls = tollRoutes.count > 0
+
+        // pick a random tail direction to keep things varied
+        guard let randomTailPosition = [RouteDurationAnnotationTailPosition.left, RouteDurationAnnotationTailPosition.right].randomElement() else { return }
+
+        var features = [MGLPointFeature]()
+
+        // Run through our heuristic algorithm looking for a good coordinate along each route line to place it's route annotation
+        // First, we will look for a set of RouteSteps unique to each route
+        var excludedSteps = [RouteStep]()
+        for (index, route) in routes.enumerated() {
+            let allSteps = route.legs.compactMap { return $0.steps }.reduce([], +)
+            let alternateSteps = allSteps.filter { step -> Bool in
+                for existingStep in excludedSteps {
+                    if step == existingStep {
+                        return false
+                    }
+                }
+                return true
+            }
+
+            excludedSteps.append(contentsOf: alternateSteps)
+            let visibleAlternateSteps = alternateSteps.filter { $0.intersects(visibleBoundingBox) }
+
+            var coordinate = kCLLocationCoordinate2DInvalid
+
+            // Obtain a polyline of the set of steps. We'll look for a good spot along this line to place the annotation.
+            // We will consider a good spot to be somewhere near the middle of the line, making sure that the coordinate is visible on-screen.
+            if let continuousLine = visibleAlternateSteps.continuousShape(), continuousLine.coordinates.count > 0 {
+                coordinate = continuousLine.coordinates[0]
+
+                // Pick a coordinate using some randomness in order to give visual variety.
+                // Take care to snap that coordinate to one that lays on the original route line.
+                // If the chosen snapped coordinate is not visible onthe screen, then we walk back along the route coordinates looking for one that is.
+                // If none of the earlier points are on screen then we walk forward along the route coordinates until we find one that is.
+                if let distance = continuousLine.distance(), let sampleCoordinate = continuousLine.indexedCoordinateFromStart(distance: distance * CLLocationDistance.random(in: 0.3...0.8))?.coordinate, let routeShape = route.shape, let snappedCoordinate = routeShape.closestCoordinate(to: sampleCoordinate) {
```

Comment:
What’s the significance of the start and end ranges of this random distance? Should it depend on the zoom level or the style? Does the range hold up better in some kinds of routes versus others?

I’m still concerned that, by focusing on geographic coordinates rather than screen coordinates as suggested in https://github.com/mapbox/mapbox-navigation-ios/pull/2734#discussion_r521687903, the whole algorithm is vulnerable to a variety of factors that are hard to control for at this high level of the stack.

In particular, `LineString.indexedCoordinateFromStart(distance:)` can only return one of the route line’s vertices, but there’s no guarantee about the density of vertices along the route line. For example, in some places in OpenStreetMap, a vertex may occur only once every quarter mile at an intersection, while in other places, it may occur every few car lengths because turn lanes and parking restrictions have been mapped in fine detail.

Reply:
The intent is to pick a coordinate along the line some distance away from the start and end points of the route, but with some variation of the distance. The goal is to not hard code something like halfway through the route.

I went with the indexed coordinate in order to reduce the computation burden of calculating a point along the line somewhere between two vertices. It might be that the overall algorithm is simplified enough that we wouldn't need to be calculating as many points and it would be worth it.

In my testing, the annotations generally work out very well, but do sometimes collide. I suspect that the code required to calculate things in screen space so that they would never collide would be significantly more complicated than just doing the geographic calculations.

My hope is that the logic we have is good enough for most of the time and that any customer who found the collision avoidance to not work to their liking would just figure out what works best for them.

Reply:
At a glance, probably the biggest risk at the moment would be that the map happens to be rotated or tilted differently than the positioning code was designed for. As long as these degrees of freedom are possible, it’ll remain difficult to completely account for them in geographic coordinate space. But to the extent that this poses a problem, we can treat it as tail work, perhaps awaiting new map SDK functionality to make such calculations easier.